app/models/internal/impl/DefaultRenderedReport.java

```diff
@@ -0,0 +1,165 @@
+        return Arrays.equals(_bytes, that._bytes)
+                && _scheduledFor.equals(that._scheduledFor)
+                && _generatedAt.equals(that._generatedAt)
+                && _format.equals(that._format);
+    }
+
+    @Override
+    public int hashCode() {
+        int result = Objects.hash(_scheduledFor, _generatedAt, _format);
+        result = 31 * result + Arrays.hashCode(_bytes);
+        return result;
+    }
+
+    @Override
+    public String toString() {
+        return MoreObjects.toStringHelper(this)
+                .add("_bytes", _bytes)
+                .add("_scheduledFor", _scheduledFor)
+                .add("_generatedAt", _generatedAt)
+                .add("_format", _format)
+                .toString();
+    }
+
+    private final byte[] _bytes;
+    private final Instant _scheduledFor;
+    private final Instant _generatedAt;
+    private final ReportFormat _format;
+
+    /**
+     * Builder implementation that constructs {@code DefaultReport}.
+     */
+    public static final class Builder extends OvalBuilder<DefaultRenderedReport> {
+        /**
+         * Public Constructor.
+         */
+        public Builder() {
+            super(DefaultRenderedReport::new);
+        }
+
+        /**
+         * Set the report bytes. Required. Cannot be null.
+         *
+         * @param bytes The report bytes.
+         * @return This instance of {@code Builder}.
+         */
+        public Builder setBytes(final byte[] bytes) {
+            _bytes = bytes.clone();
+            return this;
+        }
+
+        /**
+         * Set the report bytes. Required. Cannot be null.
+         *
+         * @param scheduledFor The report scheduledFor.
+         * @return This instance of {@code Builder}.
+         */
+        public Builder setScheduledFor(final Instant scheduledFor) {
+            _scheduledFor = scheduledFor;
+            return this;
+        }
+
+        /**
+         * Set the report generatedAt. Required. Cannot be null.
+         *
+         * @param generatedAt The report generatedAt.
+         * @return This instance of {@code Builder}.
+         */
+        public Builder setGeneratedAt(final Instant generatedAt) {
+            _generatedAt = generatedAt;
+            return this;
+        }
+
+        /**
+         * Set the report format. Required. Cannot be null.
+         *
+         * @param format The report format.
+         * @return This instance of {@code Builder}.
+         */
+        public Builder setFormat(final ReportFormat format) {
+            _format = format;
+            return this;
+        }
+
+        @NotNull
+        private byte[] _bytes;
+        @NotNull
+        private Instant _scheduledFor;
+        @NotNull
+        private Instant _generatedAt;
+        @NotNull
+        private ReportFormat _format;
+    }
+
+    private DefaultRenderedReport(final Builder builder) {
+        _bytes = builder._bytes;
```

Comment:
Using `byte[]` has problems because it is mutable. 

There are no good answers to this:

https://stackoverflow.com/questions/5480070/immutable-container-for-a-byte-that-supports-subsequences-like-string-substri

However, it's something we should at least document clearly, or if possible consider available alternatives. Definitely not suggesting solving the underlying problem entirely.